testing/unstable_stub_test.ts

```diff
@@ -0,0 +1,394 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+import { assertEquals, assertThrows } from "@std/assert";
+import {
+  assertSpyCall,
+  assertSpyCallArg,
+  assertSpyCalls,
+  MockError,
+  returnsNext,
+} from "./mock.ts";
+import { Point, type PointWithExtra } from "./_test_utils.ts";
+import { type Stub, stub } from "./unstable_stub.ts";
+
+Deno.test("stub()", () => {
+  const point = new Point(2, 3);
+  const func = stub(point, "action");
+
+  assertSpyCalls(func, 0);
+
+  assertEquals(func.call(point), undefined);
+  assertSpyCall(func, 0, {
+    self: point,
+    args: [],
+    returned: undefined,
+  });
+  assertSpyCalls(func, 1);
+
+  assertEquals(point.action(), undefined);
+  assertSpyCall(func, 1, {
+    self: point,
+    args: [],
+    returned: undefined,
+  });
+  assertSpyCalls(func, 2);
+
+  assertEquals(func.original, Point.prototype.action);
+  assertEquals(point.action, func);
+
+  assertEquals(func.restored, false);
+  func.restore();
+  assertEquals(func.restored, true);
+  assertEquals(point.action, Point.prototype.action);
+  assertThrows(
+    () => func.restore(),
+    MockError,
+    "Cannot restore: instance method already restored",
+  );
+  assertEquals(func.restored, true);
+});
+
+Deno.test("stub() works on function", () => {
+  const point = new Point(2, 3);
+  const returns = [1, "b", 2, "d"];
+  const func = stub(point, "action", () => returns.shift());
+
+  assertSpyCalls(func, 0);
+
+  assertEquals(func.call(point), 1);
+  assertSpyCall(func, 0, {
+    self: point,
+    args: [],
+    returned: 1,
+  });
+  assertSpyCalls(func, 1);
+
+  assertEquals(point.action(), "b");
+  assertSpyCall(func, 1, {
+    self: point,
+    args: [],
+    returned: "b",
+  });
+  assertSpyCalls(func, 2);
+
+  assertEquals(func.original, Point.prototype.action);
+  assertEquals(point.action, func);
+
+  assertEquals(func.restored, false);
+  func.restore();
+  assertEquals(func.restored, true);
+  assertEquals(point.action, Point.prototype.action);
+  assertThrows(
+    () => func.restore(),
+    MockError,
+    "Cannot restore: instance method already restored",
+  );
+  assertEquals(func.restored, true);
+});
+Deno.test("stub() works on getter", () => {
+  const point = new Point(5, 6);
+  const returns = [1, 2, 3, 4];
+  const func = stub(point, "x", {
+    get: function () {
+      return returns.shift();
+    },
+  });
+
+  assertSpyCalls(func.get, 0);
+
+  assertEquals(point.x, 1);
+  assertSpyCall(func.get, 0, {
+    self: point,
+    args: [],
+    returned: 1,
+  });
+  assertSpyCalls(func.get, 1);
+
+  assertEquals(point.x, 2);
+  assertSpyCall(func.get, 1, {
+    self: point,
+    args: [],
+    returned: 2,
+  });
+  assertSpyCalls(func.get, 2);
+
+  assertEquals(func.restored, false);
+  func.restore();
+  assertEquals(func.restored, true);
+  assertEquals(point.x, 5);
+  assertThrows(
+    () => func.restore(),
+    MockError,
+    "Cannot restore: instance method already restored",
```

Comment:
One minor comment, this says an instance method was restored when it was actually a property in this case. We could probably determine which it is by looking at the type of the value. Also I'd suggest renaming func to prop in this test case since it is stubbing a property rather than a func.